aten/src/ATen/native/Dropout.cpp

```diff
@@ -82,13 +83,37 @@ ALIAS_SPECIALIZATION(_feature_alpha_dropout, true,  true )
 
 } // anomymous namepsace
 
+std::tuple<Tensor,Tensor>
+native_dropout_cpu(const Tensor& input, double p, double scale, bool train) {
+  TORCH_CHECK(train, "Train parameter is incorrectly set!");
+  if (input.numel() == 0) {
+    return std::make_tuple(input, at::empty_like(input, input.options()));
+  }
+
+  auto noise = at::empty_like(input, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
+  noise.bernoulli_(p);
+
+  auto output = input.mul(noise).mul_(scale);
+  return std::make_tuple(output, noise);
+}
+
+Tensor native_dropout_backward_cpu(const Tensor& grad, const Tensor& mask, double scale) {
+  Tensor result = grad * mask * scale;
+  return result;
+}
+
 Tensor dropout(const Tensor& input, double p, bool train) {
+  TORCH_CHECK(p >= 0 && p <= 1, "dropout probability has to be between 0 and 1, but got ", p);
   auto result = [&]() {
     NoNamesGuard guard;
-    if (train && is_fused_kernel_acceptable(input, p)) {
-      return std::get<0>(at::_fused_dropout(input, 1 - p));
+    double p1m = 1. - p;
+    // Check for probability of zero to avoid divide by zero and NaN results
+    double scale = p1m == 0 ? 0. : 1. / p1m;
```

Comment:
nitpick. We are removing the short-cut and launching no-op kernels for special case p1m (0.0 & 1.0)

Reply:
I actually preferred removing the shortcuts.  I thought it makes things simpler without conditionals.  Someone who is adding the dropout operator is not using it for cases 0. and 1., except for debug use with case 1. to remove randomness, therefore, special casing them for performance did not make sense to me beyond making sure that case 0. does not cause a divide-by-zero fault.